### PR TITLE
Dl issue 182

### PR DESCRIPTION
**Pull request by dmleontiev9000**
Tempesta parser was cut from main source tree and supporting code was refactored for testing in standalone environment. Tempesta parser was patched with AVX insertions in performance critical places, but otherwise remains intact and can be transferred back to it's original position.

### CODE REVIEW

File: tempesta_fw/avx_routines.h [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    /*combine mask2 and mask4*/\
+    mask2 = _mm256_and_si256(mask2, _mm256_blend_epi32(mask4, 0xf3);\
+    bm |= 0x05040302 & _mm256_movemask_epi8(mask1);\
+    bm |= 0x09080706 & _mm256_movemask_epi8(mask2);\
+    bm = bm | (bm>>16);\
+    bm = bm | (bm>>8);\
+    result = 0xFF & bm;} while(0);
+
+#define AVX_STRID_1(n) do {\
+    static const int ids[9] = {\
```
Comment (milabs) [inline tempesta_fw/avx_routines.h]: There must be `ids[10]` here.

File: tempesta_fw/avx_routines.h [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        TFW_HTTP_HDR_CONTENT_TYPE,\
+        TFW_HTTP_HDR_RAW,\
+        TFW_HTTP_HDR_X_FORWARDED_FOR,\
+        TFW_HTTP_HDR_CONTENT_LENGTH,\
+        TFW_HTTP_HDR_RAW,\
+        TFW_HTTP_HDR_COOKIE,\
+        TFW_HTTP_HDR_CONNECTION,\
+        TFW_HTTP_HDR_USER_AGENT,\
+    };\
+    BUG_ON(n < 0 || n >= 9);\
```
Comment (milabs) [inline tempesta_fw/avx_routines.h]: Why not to use `ARRAY_SIZE(ids)` here?

File: tempesta_fw/avx_routines.h [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    return ids[n];}
+
+#define AVX_STRID_2(n) do {\
+    static const int ids[9] = {\
```
Comment (milabs) [inline tempesta_fw/avx_routines.h]: The same as above. Need to use `ids[10]` or `ids[]` as there are 10 elements in array.

File: tempesta_fw/http_parser.c [hunk 4]
```diff
@@ -86,9 +87,30 @@ enum {
...
+    };
+    k = __builtin_ctz(k);
+    s += k;
+    n -= k;
+#endif
 	while (n) {
-		if (IS_CR_OR_LF(*s))
-			return (unsigned char *)s;
+        unsigned char c = *s;
+        if (IS_CR_OR_LF(c))
```
Comment (milabs) [inline tempesta_fw/http_parser.c]: No need to change the code. Your change doesn't improve it.

File: tempesta_fw/http_parser.c [hunk 5]
```diff
@@ -1809,11 +1834,73 @@ tfw_http_parse_req(void *req_data, unsigned char *data, size_t len)
 	__FSM_STATE(RGen_Hdr) {
 		tfw_http_msg_hdr_open(msg, p);
 
-		if (unlikely(IS_CR_OR_LF((c))))
-			__FSM_JMP(RGen_EoL);
-
-		if (unlikely(!IN_ALPHABET(c, hdr_a)))
-			return TFW_BLOCK;
+        if (unlikely(IS_CR_OR_LF((c))))
```
Comment (milabs) [inline tempesta_fw/http_parser.c]: Wha's wrong with the indention?

File: tempesta_fw/http_parser.c [hunk 6]
```diff
@@ -1809,11 +1834,73 @@ tfw_http_parse_req(void *req_data, unsigned char *data, size_t len)
...
+            __FSM_JMP(RGen_EoL);
+
+        if (unlikely(!IN_ALPHABET(c, hdr_a)))
```
Comment (milabs) [inline tempesta_fw/http_parser.c]: The same as above.

File: tempesta_fw/avx_routines.h [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    unsigned char __avx_realign[64] __attribute__((aligned(64))); \
+    do {\
+        __m256 tmp = _mm256_load(__c_avx_data);\
+        __m256 tmp2 = _mm256_load(__c_avx_data+32);\
+        HEADER_CHARSET = _mm256_permute2f128_si256(tmp, tmp, 0x00);\
+        __m256 tmp3 = _mm256_permute4x64_epi64(tmp, 0xEE);\
+        tmp2 = _mm256_unpacklo_epi32(tmp3, tmp3);\
+        __c_avx_constants1 = tmp3;\
+        __c_avx_constants2 = tmp2;\
+    }while(0);
```
Comment (milabs) [inline tempesta_fw/avx_routines.h]: As I see it, `HEADER_CHARSET, __c_avx_constants1, __c_avx_constants2` can be precalculated.

File: tempesta_fw/avx_routines.h [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+#define __mm_align(value, n) do { \
+    _mm256_store_si256((__m256i*)__avx_realign, value);\
+    _mm256_store_si256((__m256i*)(__avx_realign+32, _mm256_setzero_si256()));\
+    value = _mm256_loadu_si256((__m256i*)(__avx_realign+n));\
+    }while(0);
+
+enum {
```
Comment (milabs) [inline tempesta_fw/avx_routines.h]: Not sure, that `enum` is the right method of declaring such a constants.

File: tempesta_fw/avx_routines.h (hunk 1 above)
Comment (dmleontiev9000) [inline tempesta_fw/avx_routines.h]: Fixed in latest commit

2016-06-27 13:01 GMT+04:00 Ilya V. Matveychikov <notifications@github.com>:

File: tempesta_fw/avx_routines.h (hunk 2 above)
Comment (dmleontiev9000) [inline tempesta_fw/avx_routines.h]: Fixed in latest commit

File: tempesta_fw/avx_routines.h (hunk 3 above)
Comment (dmleontiev9000) [inline tempesta_fw/avx_routines.h]: Fixed in latest commit

**Comment by krizhanovsky** [comment]
The code doesn't comply our [coding style](https://github.com/tempesta-tech/linux-4.1-tfw/blob/master/Documentation/CodingStyle), please adjust the style.

**Comment by krizhanovsky** [comment]
What's going on in `tempesta_fw/t/unit/dl-182`? It seems there should be test/benchmark for HTTP parser, but why do you copy skb stuff there? Please look at other unit tests.

File: tempesta_fw/t/unit/dl-182/main.c [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    FILE * source = fopen(argv[nfile], "rt");
+    if (!source) {
+        printf("error: input file '%s' cannot be opened\n", argv[nfile]);
+        return 1;
+    }
+
+    fseek(source, 0, SEEK_END);
+    size_t len = ftell(source);
```
Comment (krizhanovsky) [inline tempesta_fw/t/unit/dl-182/main.c]: usually `fstat(2)` is used to get file size. Moreover, since you set file pointer using `fseek()` following `fread()` doesn't read any data and your test doesn't do anything.

File: tempesta_fw/avx_routines.c [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    0xFF,0xFF,0xFF,0xFF,0xFF,0x00,0x00,0x00,
+    0xFF,0xFF,0xFF,0x00,0x00,0x00,0x00,0x00,
+    0xFF,0xFF,0xFF,0x00,0x00,0x00,0x00,0x00,
+    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
+
+    'n','t','r','o','l', 0 , 0 , 0 ,
+    'g', 0 , 0 , 0 , 0 , 0 , 0 , 0 ,
+    'o','n', 0 , 0 , 0 , 0 , 0 , 0 ,
+    'n','t', 0 , 0 , 0 , 0 , 0 , 0 ,
+};
```
Comment (krizhanovsky) [inline tempesta_fw/avx_routines.c]: The file contains constants only. And the constants are used in different compilation unit. Why did you move them to separate .c file? Moreover, since this is standalone translation unit, compiler can not efficiently optimize code using the constants and directly build them into code.

File: tempesta_fw/http_parser.c [hunk 11]
```diff
@@ -1809,11 +1834,73 @@ tfw_http_parse_req(void *req_data, unsigned char *data, size_t len)
...
+            //           Host:   yandex.ru\n
+            // tmp               *
+            if (unlikely(!tmp)) {
+                //we got header name and some spaces after it, but
+                //not header value at all
+                __FSM_MOVE_n(RGen_LWS, 32);
+            }
+            __FSM_MOVE_n(RGen_LWS, __builtin_ctz(tmp));
+        }
+#endif
```
Comment (krizhanovsky) [inline tempesta_fw/http_parser.c]: This is not subject for #182. Meantime, `memchr()` which ARE subject of #182 still exist in code, so 'other' headers aren't verified for allowed alphabet.

File: tempesta_fw/http_parser.c [hunk 12]
```diff
@@ -86,9 +87,30 @@ enum {
 static inline unsigned char *
 memchreol(const unsigned char *s, size_t n)
 {
+#ifdef __AVX2__
+    __m256 __avx_constants2 = __constants2;
```
Comment (krizhanovsky) [inline tempesta_fw/http_parser.c]: `__constants2` aren't defined in `tempesta_fw` tree, but in `./t/unit/dl-182/avx_routines.h` only.

**Comment by krizhanovsky** [comment]
Poorly done
